src/clients/UBAClient/UBAClientUtilities.ts

```diff
@@ -77,3 +127,336 @@ export async function getUBAFeeConfig(
     }
   );
 }
+
+export async function updateUBAClient(
+  hubPoolClient: HubPoolClient,
+  spokePoolClients: { [chainId: number]: SpokePoolClient },
+  relevantChainIds: number[],
+  relevantTokenSymbols: string[],
+  hubPoolBlockNumber: number,
+  updateInternalClients = true,
+  relayFeeCalculatorConfig: RelayFeeCalculatorConfig,
+  currentState?: {
+    [chainId: number]: UBAChainState;
+  }
+): Promise<{
+  [chainId: number]: UBAChainState;
+}> {
+  if (updateInternalClients) {
+    await hubPoolClient.update();
+    await Promise.all(Object.values(spokePoolClients).map((spokePoolClient) => spokePoolClient.update()));
+  }
+  const ubaChainStates: { [chainId: number]: UBAChainState } = {};
+
+  for (const chainId of relevantChainIds) {
+    const spokePoolClient = spokePoolClients[chainId];
+
+    const chainState: UBAChainState = {
+      bundles: currentState?.[chainId]?.bundles ?? {},
+      spokeChain: {
+        deploymentBlockNumber: spokePoolClient.deploymentBlock,
+        bundleEndBlockNumber: hubPoolClient.getLatestBundleEndBlockForChain(
+          relevantChainIds,
+          hubPoolBlockNumber,
+          chainId
+        ),
+        latestBlockNumber: spokePoolClient.latestBlockNumber,
+      },
+    };
+
+    for (const tokenSymbol of relevantTokenSymbols) {
+      // Get the bundles for this token
+      const availableBundles = chainState.bundles[tokenSymbol] ?? [];
+      // Get the block number and opening balance for this token
+      const { blockNumber, spokePoolBalance, incentiveBalance } = getOpeningTokenBalances(
```

Comment:
This `blockNumber` is the `bundleEndBlock` for the `chainId` right? I think it might be helpful to rename so as not to confuse yourself about which chain you're comparing blocks on

Reply:
This new PR regenerates all the bundles on update. https://github.com/across-protocol/sdk-v2/pull/266/commits/617c110861852a75cfe6c538ac56c5e4e0901f03